include/exadg/incompressible_navier_stokes/spatial_discretization/generalized_newtonian_model.h

```diff
@@ -0,0 +1,164 @@
+   */
+  void
+  set_viscosity(VectorType const & velocity) const;
+
+  /*
+   *  Function for *adding to* the viscosity taking the currently stored viscosity as a basis.
+   */
+  void
+  add_viscosity(VectorType const & velocity) const;
+
+private:
+  void
+  cell_loop_set_coefficients(dealii::MatrixFree<dim, Number> const & data,
+                             VectorType &,
+                             VectorType const & src,
+                             Range const &      cell_range) const;
+
+  void
+  face_loop_set_coefficients(dealii::MatrixFree<dim, Number> const & data,
+                             VectorType &,
+                             VectorType const & src,
+                             Range const &      face_range) const;
+
+  void
+  boundary_face_loop_set_coefficients(dealii::MatrixFree<dim, Number> const & data,
+                                      VectorType &,
+                                      VectorType const & src,
+                                      Range const &      face_range) const;
+
+  /*
+   *  This function computes the kinematic viscosity for
+   *  generalized Newtonian fluids, i.e., based on the shear rate.
+   *  All models can be found in , e.g., Galdi et al., 2008
+   *  ("Hemodynamical Flows: Modeling, Analysis and Simulation").
+   *  With
+   *  y    = sqrt(2*sym_grad_velocity : sym_grad_velocity)
+   *  e_oo = generalized_newtonian_kinematic_viscosity_lower_limit
+   *  e_0  = generalized_newtonian_kinematic_viscosity_upper_limit
+   *  k    = generalized_newtonian_kappa
+   *  l    = generalized_newtonian_lambda
+   *  a    = generalized_newtonian_a
+   *  b    = generalized_newtonian_b
+   *  we have the apparent viscosity
+   *  viscosity = e_oo + (e_0 - e_oo) * [k + (l * y)^a]^[(n - 1) / a]
+   *
+   *  We distinguish the cases:
+   *  GeneralizedCarreauYasuda, // no assumptions
+   *  Carreau,                  // k = 1, a = 2
+   *  Cross,                    // k = 1, n = 1 - a
+   *  SimplifiedCross,          // k = 1, a = 1, n = 0
+   *  PowerLaw                  // k = 0 (e_00 for numerical reasons)
+   */
+  void
+  add_generalized_newtonian_viscosity(scalar & viscosity, tensor const & velocity_gradient) const;
+
+  void
+  compute_viscosity_factor(scalar & viscosity_factor, scalar const & shear_rate) const;
+
+  /*
+   * The rheological models below differ in the factor scaling the margin, (e_0 - e_oo),
+   * and hence only compute the factor [k + (l * y)^a]^[(n - 1) / a] .
+   */
+  void
+  generalized_carreau_yasuda_model(scalar & viscosity_factor, scalar const & shear_rate) const;
+
+  void
+  carreau_model(scalar & viscosity_factor, scalar const & shear_rate) const;
+
+  void
+  cross_model(scalar & viscosity_factor, scalar const & shear_rate) const;
+
+  void
+  simplified_cross_model(scalar & viscosity_factor, scalar const & shear_rate) const;
+
+  void
+  power_law_model(scalar & viscosity_factor, scalar const & shear_rate) const;
```

Comment:
As written here https://github.com/exadg/exadg/pull/381/files#r1172220169, I think there is not need for implementing these functions. I think this will really simplify the code overall.